File: apps/desktop/src/hooks/pluginRuntimeCommands/wall.ts
import { invoke } from "@tauri-apps/api/core";
import { pluginWallLayouts, pluginWallTiles } from "../../app/pluginRuntime";
import type { MpvWallTileSnapshot } from "../../app/types";
import { PLUGIN_RUNTIME_COMMAND_NOT_HANDLED, type PluginRuntimeCommandHandler } from "./types";

export const handlePluginWallRuntimeCommand: PluginRuntimeCommandHandler = async (context, command, record, permissions) => {
  switch (command) {
    case "player.wall.open": {
      if (!permissions.has("mpv.wall")) {
        throw new Error("plugin runtime command requires mpv.wall");
      }
      return await invoke<MpvWallTileSnapshot[]>("mpv_wall_open", { tiles: pluginWallTiles(record.tiles, context.pluginViewFrameRef.current) });
    }
    case "player.wall.layout": {
      if (!permissions.has("mpv.wall")) {
        throw new Error("plugin runtime command requires mpv.wall");
      }
      return await invoke<MpvWallTileSnapshot[]>("mpv_wall_layout", { tiles: pluginWallLayouts(record.tiles, context.pluginViewFrameRef.current) });
    }
    case "player.wall.snapshot": {
      if (!permissions.has("mpv.wall")) {
        throw new Error("plugin runtime command requires mpv.wall");
      }
      return await invoke<MpvWallTileSnapshot[]>("mpv_wall_snapshot");
    }
    case "player.wall.setVisible": {
      if (!permissions.has("mpv.wall")) {
        throw new Error("plugin runtime command requires mpv.wall");
      }
      await invoke("mpv_wall_set_visible", { visible: record.visible !== false });
      return null;
    }
    case "player.wall.close": {
      if (!permissions.has("mpv.wall")) {
        throw new Error("plugin runtime command requires mpv.wall");
      }
      await invoke("mpv_wall_close");
      return null;
    }
    default:
      return PLUGIN_RUNTIME_COMMAND_NOT_HANDLED;
  }
};

File: apps/desktop/src-tauri/src/mpv_embed/commands/wall.rs
use super::super::*;

#[tauri::command]
pub async fn mpv_wall_open(
    app: AppHandle,
    tiles: Vec<MpvWallTileRequest>,
) -> Result<Vec<MpvWallTileSnapshot>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let state = app.state::<MpvWallState>();
        wall_open_for_app(&app, state.inner(), tiles)
    })
    .await
    .map_err(|error| format!("mpv wall command task failed: {error}"))?
}

#[tauri::command]
pub async fn mpv_wall_layout(
    app: AppHandle,
    tiles: Vec<MpvWallTileLayout>,
) -> Result<Vec<MpvWallTileSnapshot>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let state = app.state::<MpvWallState>();
        wall_layout_for_app(&app, state.inner(), tiles)
    })
    .await
    .map_err(|error| format!("mpv wall command task failed: {error}"))?
}

#[tauri::command]
pub async fn mpv_wall_snapshot(app: AppHandle) -> Result<Vec<MpvWallTileSnapshot>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let state = app.state::<MpvWallState>();
        wall_snapshot(state.inner())
    })
    .await
    .map_err(|error| format!("mpv wall command task failed: {error}"))?
}

#[tauri::command]
pub async fn mpv_wall_close(app: AppHandle) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        let state = app.state::<MpvWallState>();
        wall_close(&app, state.inner())
    })
    .await
    .map_err(|error| format!("mpv wall command task failed: {error}"))?
}

#[tauri::command]
pub async fn mpv_wall_set_visible(app: AppHandle, visible: bool) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        let state = app.state::<MpvWallState>();
        wall_set_visible(&app, state.inner(), visible)
    })
    .await
    .map_err(|error| format!("mpv wall command task failed: {error}"))?
}
